### PR TITLE
10 add llm as judge metric

### LINKED ISSUE #10: add llm as judge metric
Add option for metric that uses llm as judge e.g. gptscore

### PR DESCRIPTION
**Pull request by avishaiElmakies**
Adds llm as judge metric to the library.
Use `metric=llm_as_judge`
As the judge you can use either an llm from hf or an openAI model using thier api.
If using openai api set env variable `OPENAI_API_KEY` to the key for open ai api.
also add option to filter prompts shorter than `min_file_length`
add option to use text alignment to cut the prompt so it doesn't cut a word. use `use_alignment=true` and if you're alignment is in a different folder use `alignment_folder=<>`

### CODE REVIEW

File: config/metric/asr_perplexity.yaml [hunk 1]
```diff
@@ -14,3 +14,4 @@ generate_kwargs:
   top_k: 25
   max_new_tokens: 150
   do_sample: true
+  repetition_penalty: 1.1
```
Comment (gallilmaimon) [inline config/metric/asr_perplexity.yaml]: I think we should add a comment about this here like we have in generate explaining what this solves. Also I wonder if it should be the default of the metric, as non-DPO models don't use this

File: config/metric/llm_as_judge.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+min_file_length: 6
+use_alignment: true
+alignment_folder: null                
+num_log: 10
+generate_kwargs:
+  temperature: 0.8
+  top_k: 25
+  max_new_tokens: 150
+  do_sample: true
+  repetition_penalty: 1.1
```
Comment (gallilmaimon) [inline config/metric/llm_as_judge.yaml]: See my other comment

File: cli/eval.py [hunk 3]
```diff
@@ -94,6 +106,10 @@ def main(cfg: DictConfig):
                                                                caption=f'generated_{i}', sample_rate=tokeniser.fe_sample_rate)
                 logs[f'prompt/prompt_{i}'] = wandb.Audio(prompt.squeeze(0).cpu().numpy(),
                                                          caption=f'prompt_{i}', sample_rate=tokeniser.fe_sample_rate)
+                if "text_res" in res:
```
Comment (gallilmaimon) [inline cli/eval.py]: Maybe we should rename text_res to something more specific? I assume this is the transcriptions of the audio input and output and not a text response (in a multi-modal setting for instance)

File: cli/eval.py [hunk 4]
```diff
@@ -64,7 +71,12 @@ def main(cfg: DictConfig):
         for key, val in res.items():
             if key == "generate" or key == "prompts":
                 continue
-            print(f"{key}: {val}")
+            if isinstance(val, list):
+                print(f"{key}:")
```
Comment (gallilmaimon) [inline cli/eval.py:75]: What are these prints? maybe they should be logs and with slightly more context if they are needed?

File: slamkit/metric/generative_metric.py [hunk 5]
```diff
@@ -6,24 +7,49 @@
...
+import os
 
-from .metric_utils import get_whisper_pipeline, get_llm, get_llm_preplexity
+from .metric_utils import get_whisper_pipeline, get_llm, get_llm_preplexity, get_judge
 from ..model.speech_lm import SpeechLM
 from ..utils.calculation_utils import calc_auto_bleu
 
 logger = logging.getLogger(__name__)
 
+def get_cut_location(alignment, prompt_length):
```
Comment (gallilmaimon) [inline slamkit/metric/generative_metric.py]: Perhaps add typing, explain the format of the alignment briefly and say if the prompt length is a float in seconds?

File: slamkit/metric/generative_metric.py [hunk 6]
```diff
@@ -6,24 +7,49 @@
...
+    """
+    Find the closest word in the alignment to the prompt length.
```
Comment (gallilmaimon) [inline slamkit/metric/generative_metric.py]: This actually finds the time which is the word end which is closest, not the word itself

File: slamkit/metric/generative_metric.py [hunk 7]
```diff
@@ -6,24 +7,49 @@
...
+    """
+    endtimes = torch.tensor([word[2] for word in alignment])
+    dist = torch.abs(endtimes - prompt_length)
+    time = endtimes[dist.argmin()].item()        
+    return time
 
+def is_shorter(file, min_file_length):
```
Comment (gallilmaimon) [inline slamkit/metric/generative_metric.py]: Types + units (e.g float, seconds)

File: slamkit/metric/generative_metric.py [hunk 8]
```diff
@@ -6,24 +7,49 @@
...
         self.sample_rate = sample_rate
         if num_files is None:
             self.data = glob(glob_path, recursive=True)
+            if min_file_length is not None:
+                self.data = list(filter(lambda x: not is_shorter(x, min_file_length), self.data))
         else:
+            self.data = []
             paths = iglob(glob_path, recursive=True)
-            self.data = [next(paths) for _ in range(num_files)]
+            for path in paths:
```
Comment (gallilmaimon) [inline slamkit/metric/generative_metric.py:45]: Maybe we want to catch or deal with StopIterationn here if the glob has only n files and you asked for m > n?

File: slamkit/metric/generative_metric.py [hunk 9]
```diff
@@ -87,3 +132,41 @@ def asr_perplexity(model: SpeechLM, data_path: str, batch_size: int, whisper_mod
...
+    res, gen, prompts, texts = [], [], [], []
+    prompt_texts, gen_texts = [], []
+    for audio, l in tqdm(dl, desc="Generating"):
+        audio = audio.to(model.device)
+        gen_res = model.generate(audio, l, used_tokens_modality, remove_prompt=True, **generate_kwargs)
+        gen.extend(gen_res)
+        prompts.extend(audio)
+        prompt_audio_transcriptions = whisper_pipeline([sample.cpu().numpy() for sample in audio], batch_size=len(audio))
+        generated_audio_transcriptions = whisper_pipeline([gen.cpu().numpy() for gen in gen_res], batch_size=len(gen_res))
+        prompt_audio_transcriptions = [res_text["text"] if audio.shape[-1] > 0 else "" for audio, res_text in zip(audio, prompt_audio_transcriptions)]
```
Comment (gallilmaimon) [inline slamkit/metric/generative_metric.py]: Also here not sure about the name res_text

File: slamkit/metric/generative_metric.py [hunk 10]
```diff
@@ -87,3 +132,41 @@ def asr_perplexity(model: SpeechLM, data_path: str, batch_size: int, whisper_mod
...
+        generated_audio_transcriptions = [res_text["text"] if gen.shape[-1] > 0 else "" for gen, res_text in zip(gen_res, generated_audio_transcriptions)]
+        text = [
```
Comment (gallilmaimon) [inline slamkit/metric/generative_metric.py:162]: Maybe there is a cleaner way to this with a dictionary of mappings or a utility?

File: slamkit/metric/generative_metric.py [hunk 11]
```diff
@@ -87,3 +132,41 @@ def asr_perplexity(model: SpeechLM, data_path: str, batch_size: int, whisper_mod
...
+            instruction.replace("[prompt_audio_transcription]", prompt_audio_transcription).replace("[generated_audio_transcription]", generated_audio_transcription) for 
+            prompt_audio_transcription, generated_audio_transcription in zip(prompt_audio_transcriptions, generated_audio_transcriptions)
+        ]
+        texts.extend(text)
+        prompt_texts.extend(prompt_audio_transcriptions)
+        gen_texts.extend(generated_audio_transcriptions)
+    res = judge(texts)
+    res = list(filter(lambda x: x is not None, res))
+    text_res = [(p_text,g_text) for p_text, g_text, r in zip(prompt_texts, gen_texts, res)]
+    print("got response for {} out of {}".format(len(res), len(dataset)))
```
Comment (gallilmaimon) [inline slamkit/metric/generative_metric.py]: Why printed? Either logged or not at all, no?

File: slamkit/metric/metric_utils.py [hunk 12]
```diff
@@ -47,3 +58,73 @@ def get_llm_preplexity(model:PreTrainedModel, tokeniser:PreTrainedTokenizer, tex
...
+    return None
+
+def judge_text(model, tokenizer, text:List[str], device):
+    
+    tokenizer.padding_side = "left"
+    model_inputs = tokenizer(text, return_tensors='pt', padding=True).to(device)
+    generation = model.generate(
+        input_ids=model_inputs['input_ids'],
+        attention_mask=model_inputs['attention_mask'],
+        max_new_tokens=512,
```
Comment (gallilmaimon) [inline slamkit/metric/metric_utils.py:74]: Do we want these constants defined here? Maybe configured for the metric? Not sure

File: slamkit/metric/metric_utils.py [hunk 13]
```diff
@@ -47,3 +58,73 @@ def get_llm_preplexity(model:PreTrainedModel, tokeniser:PreTrainedTokenizer, tex
...
     return calc_nll(shift_logits, shift_labels, shift_labels.ne(-100))
+
+def extract_digit_from_boxed(string):
+    match = re.search(r'\\boxed\{(\d+)\}', string)
+    if match:
+        return int(match.group(1))
+    return None
+
+def judge_text(model, tokenizer, text:List[str], device):
+
```
Comment (gallilmaimon) [inline slamkit/metric/metric_utils.py]: Remove empty line

File: slamkit/metric/metric_utils.py [hunk 14]
```diff
@@ -47,3 +58,73 @@ def get_llm_preplexity(model:PreTrainedModel, tokeniser:PreTrainedTokenizer, tex
...
     shift_labels = labels[..., 1:].contiguous()
     return calc_nll(shift_logits, shift_labels, shift_labels.ne(-100))
+
+def extract_digit_from_boxed(string):
+    match = re.search(r'\\boxed\{(\d+)\}', string)
+    if match:
+        return int(match.group(1))
+    return None
+
+def judge_text(model, tokenizer, text:List[str], device):
```
Comment (gallilmaimon) [inline slamkit/metric/metric_utils.py]: perhaps we should be consistent with tokeniser instead of tokenizer

File: slamkit/metric/metric_utils.py [hunk 15]
```diff
@@ -47,3 +58,73 @@ def get_llm_preplexity(model:PreTrainedModel, tokeniser:PreTrainedTokenizer, tex
...
+
+    def __call__(self, text:List[str]):
+        res = []
+        for i in tqdm(range(0, len(text), self.batch_size), desc="LLM Judging"):
+            batch_text = text[i:i + self.batch_size]
+            res.extend(judge_text(self.model, self.tokenizer, batch_text, self.device))
+        return res
+    
+
+class OpenAIJudge:
```
Comment (gallilmaimon) [inline slamkit/metric/metric_utils.py:97]: Do we want inheritance or some ABC for both?

File: config/metric/asr_perplexity.yaml (hunk 1 above)
Comment (avishaiElmakies) [reply config/metric/asr_perplexity.yaml]: I rember you wanted it there so i added

File: cli/eval.py (hunk 4 above)
Comment (avishaiElmakies) [reply cli/eval.py:75]: those print are to output the results from the metric

File: slamkit/metric/generative_metric.py (hunk 8 above)
Comment (avishaiElmakies) [reply slamkit/metric/generative_metric.py:45]: when you use `for` it fix the issue and stop the loop when there are less then n items

File: slamkit/metric/metric_utils.py (hunk 12 above)
Comment (avishaiElmakies) [reply slamkit/metric/metric_utils.py:74]: it could be params for config. but i think i don't want to do it now. the code is very ugly and adding more params is very annoying

File: slamkit/metric/metric_utils.py (hunk 15 above)
Comment (avishaiElmakies) [reply slamkit/metric/metric_utils.py:97]: I don't think it is needed for two classes that i don't think we will add more. if we add more then maybe

File: config/metric/asr_perplexity.yaml (hunk 1 above)
Comment (gallilmaimon) [reply config/metric/asr_perplexity.yaml]: We anyway need a comment, if it is here - to explain what it is. but I think I wouldn't put for default as part of the metric as this can be confusing. Maybe in the future we can make a separate config for generation hyper-params and then have different one for post-do and pre-do and then you can move between them easily.

File: cli/eval.py (hunk 4 above)
Comment (gallilmaimon) [reply cli/eval.py:75]: I think we should log them, not print, and either before or after give a log message to explain what they are otherwise it is a bit out of context.

File: slamkit/metric/generative_metric.py (hunk 8 above)
Comment (gallilmaimon) [reply slamkit/metric/generative_metric.py:45]: ohh, okay, right.

File: slamkit/metric/metric_utils.py (hunk 15 above)
Comment (gallilmaimon) [reply slamkit/metric/metric_utils.py:97]: Just so that they are both of the same type and other code can deduce expected output types of methods etc, but whichever you think is best I am okay with this.

File: cli/eval.py (hunk 4 above)
Comment (avishaiElmakies) [reply cli/eval.py:75]: the point of the script is to give eval. what is the problem with printing results from a metric?
and I don't know that the keys are. that is why i print them. 
this is currently how we handle all metrics. and I don't want to change that too much. 
I do want to refactor eval and fix many of the issue here. but they need more time to be fixed

File: slamkit/metric/metric_utils.py (hunk 15 above)
Comment (avishaiElmakies) [reply slamkit/metric/metric_utils.py:97]: I can add typing

File: slamkit/metric/generative_metric.py [hunk 16]
```diff
@@ -15,16 +15,17 @@
 
 logger = logging.getLogger(__name__)
 
-def get_cut_location(alignment, prompt_length):
+def get_cut_location(alignment:List[Tuple[str,float,float]], prompt_length:float):
```
Comment (gallilmaimon) [inline slamkit/metric/generative_metric.py]: I would add a comment specifying that it is time in seconds here and in the function below and add a space after : before type

File: slamkit/metric/metric_utils.py (hunk 12 above)
Comment (gallilmaimon) [reply slamkit/metric/metric_utils.py:74]: Okay for now

**Review by gallilmaimon — APPROVED** [review]
